btree/btree.c

```diff
@@ -1923,24 +1897,28 @@ static int64_t node_get(struct node_op *op, struct td *tree,
  *
  * @param op load operation to perform.
  * @param node node descriptor.
- * @param lock_acquired true if lock is already taken else false.
  * @param tx changes will be captured in this transaction.
  *
  */
-static void node_put(struct node_op *op, struct nd *node, bool lock_acquired,
-		     struct m0_be_tx *tx)
+static void node_put(struct node_op *op, struct nd *node, struct m0_be_tx *tx)
 {
 	M0_PRE(node != NULL);
 	int        shift = node->n_type->nt_shift(node);
-	segops->so_node_put(node, lock_acquired);
+	segops->so_node_put(node);
 
+	node_lock(node);
```

Comment:
With current implementation, if node_free and node_put get called at the same time with n_ref =2,now one thread mem_node_put will decrease ref counter and then other thread can take lock and decrease the n_ref and it will see n_ref as 0 and will free the node, the first thread is pointing out to the node which is not valid, which contains garbage values. 
move node_lock(node) before segops->so_node_put(node); and remove the lock taken in mem_node_put. as lock will be taken in node_put() function .

Reply:
The same thing is possible when we call node_get and node_free/ node_put for same node with ref count as 1

Reply:
Added global lock for node_* operations.